src/Dialect/ONNX/ONNXOps/ControlFlow/Loop.cpp

```diff
@@ -42,56 +47,44 @@ LogicalResult ONNXLoopOp::inferShapes(
   loopBody.getArgument(1).setType(
       RankedTensorType::get({}, builder.getI1Type()));
 
-  // Set types for loop carried dependencies (i.e., set these loop carried
-  // dependencies that appear in the body function input signature to have the
-  // same type as their counterpart in LoopOp inputs).
-  auto bodyInputs = loopBody.getArguments();
-  auto bodyVRange = llvm::make_range(bodyInputs.begin() + 2, bodyInputs.end());
-  for (auto opVToBodyVTy : llvm::zip(getVInitial(), bodyVRange)) {
-    auto opVTy = std::get<0>(opVToBodyVTy).getType();
-    std::get<1>(opVToBodyVTy).setType(opVTy);
-  }
+  // Set body input types for loop carried dependencies to the types of
+  // their LoopOp input counterpart.
+  auto bodyCarriedInputs = llvm::drop_begin(loopBody.getArguments(), 2);
+  for (auto [opInput, bodyInput] : llvm::zip(getVInitial(), bodyCarriedInputs))
+    bodyInput.setType(opInput.getType());
 
-  // Now we have modified loop body function input signatures according to
-  // the knowledge we have on the inputs we pass to this function. Dispatch
-  // shape inference to obtain body function output types.
+  // Now we have modified loop body input types according to
+  // the knowledge we have on the initial inputs. Dispatch
+  // shape inference to obtain body output types.
   doShapeInference(loopBody);
+  Operation *terminator = loopBody.back().getTerminator();
+  assert(terminator->getNumOperands() == 1 + getVFinalAndScanOutputs().size());
```

Comment:
Add the assertion error message
`&& "number of Values returned from LoopOp body and Loop output does not match"`

Reply:
changed to "LoopOp outputs count must match body results count"

also added a message in the corresponding assert in Scan.cpp